sunpy/coordinates/frames.py

```diff
@@ -847,3 +858,186 @@ class GeocentricEarthEquatorial(SunPyBaseCoordinateFrame):
     Aberration due to Earth motion is not included.
     """
     equinox = TimeFrameAttributeSunPy(default=_J2000)
+
+
+class BaseMagnetic(SunPyBaseCoordinateFrame):
+    """
+    Base class for frames that rely on the Earth's magnetic model (MAG, SM, and GSM).
+
+    This class is not intended to be used directly and has no transformations defined.
+    """
+    magnetic_model = Attribute(default='igrf13')
+
+    @property
+    def _igrf_file(self):
+        if not self.magnetic_model.startswith("igrf"):
+            raise ValueError
+
+        # First look if the file is bundled in package
+        local_file = os.path.join(os.path.dirname(__file__), "data",
+                                  f"{self.magnetic_model}coeffs.txt")
+        if os.path.exists(local_file):
+            return local_file
+
+        # Otherwise download the file and cache it
+        return cache.download("https://www.ngdc.noaa.gov/IAGA/vmod/coeffs/"
+                              f"{self.magnetic_model}coeffs.txt", expiry=None)
```

Comment:
I'll push back on switching to using Astropy's `get_pkg_data_filename()`.  We're not really using it as intended, including depending on catching the error it raises because it won't find the file on Astropy's data server.  I think the direct code is more straightforward and more readable/maintainable.

I see only one place, at least in core, where we use `get_pkg_data_filename()`, which is in `get_test_filepath()`.  Its usage there arguably makes little sense as well.

Reply:
Fair enough. I don't like using `dirname`, but using `get_pkg_data_filename` is objectively worse here.